README.md
inmetr: Historical Data from Brazilian Meteorological Stations in R
================

<!-- README.md is generated from README.Rmd. Please edit that file -->
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.580813.svg)](https://doi.org/10.5281/zenodo.580813) [![Build Status](https://travis-ci.org/lhmet/inmetr.svg?branch=master)](https://travis-ci.org/lhmet/inmetr)

Overview
--------

`inmetr` provide access to historical data measured by meteorological stations available in the Meteorological Database for Education and Research ([BDMEP](http://www.inmet.gov.br/projetos/rede/pesquisa/)) from National Institute of Meteorology (Instituto Nacional de Meteorologia - [INMET](http://www.inmet.gov.br)), Brazil.

Installation
------------

`inmetr` is easy to install from Git Hub using the `devtools` package.

``` r
library(devtools)
```

``` r
install_github('lhmet/inmetr')
```

Load package

``` r
library(inmetr)
```

Stations ID
-----------

To search a meteorological station from INMET we can use metadata of INMET stations included in `inmetr` package as `bdmep_meta`.

``` r
head(bdmep_meta)
#>      id       lon        lat    alt            name   state uf
#> 1 83010 -68.73333 -11.016667 260.00       Brasiléia    Acre AC
#> 2 82704 -72.66667  -7.633333 170.00 Cruzeiro do Sul    Acre AC
#> 3 82915 -67.80000  -9.966667 160.00      Rio Branco    Acre AC
#> 4 82807 -70.76667  -8.166667 190.00        Tarauacá    Acre AC
#> 5 83098 -36.16667 -10.150000  56.13        Coruripe Alagoas AL
#> 6 82994 -35.70000  -9.666667  64.50          Maceió Alagoas AL
#>            time_zone offset_utc        time_zone.1 offset_utc.1
#> 1 America/Rio_Branco         -5 America/Rio_Branco           -5
#> 2 America/Rio_Branco         -5 America/Rio_Branco           -5
#> 3 America/Rio_Branco         -5 America/Rio_Branco           -5
#> 4 America/Rio_Branco         -5 America/Rio_Branco           -5
#> 5     America/Maceio         -3     America/Maceio           -3
#> 6     America/Maceio         -3     America/Maceio           -3
tail(bdmep_meta)
#>        id       lon        lat    alt           name     state uf
#> 384 83033 -48.30000 -10.183333 280.00         Palmas Tocantins TO
#> 385 83231 -47.83333 -12.550000 275.00         Paranã Tocantins TO
#> 386 82863 -48.18333  -8.966667 187.00   Pedro Afonso Tocantins TO
#> 387 83228 -48.35000 -12.016667 242.49          Peixe Tocantins TO
#> 388 83064 -48.41667 -10.716667 239.20 Porto Nacional Tocantins TO
#> 389 83235 -46.41667 -12.400000 603.59     Taguatinga Tocantins TO
#>             time_zone offset_utc       time_zone.1 offset_utc.1
#> 384 America/Araguaina         -3 America/Araguaina           -3
#> 385 America/Araguaina         -3 America/Araguaina           -3
#> 386 America/Araguaina         -3 America/Araguaina           -3
#> 387 America/Araguaina         -3 America/Araguaina           -3
#> 388 America/Araguaina         -3 America/Araguaina           -3
#> 389 America/Araguaina         -3 America/Araguaina           -3
```

`bdmep_meta` is a data frame providing the `id` of stations, a numeric code defined by [OMM](http://www.wmo.int/pages/prog/www/ois/volume-a/StationIDs_Global_1509.pdf). This `id` is a necessary argument to `bdmep_import()` function which allows to download data from meteorological stations into R.

Here, we show how to find the [OMM code](http://www.wmo.int/pages/prog/www/ois/volume-a/StationIDs_Global_1509.pdf) for meteorological stations at two cities (randomly sampled).

``` r
stations <- c("Santa Maria", "Macapá")
# random sample of two stations names 
#stations <- sample(bdmep_meta$name, 2)
stations_rows <- pmatch(stations, bdmep_meta$name)
bdmep_meta[stations_rows, ]
#>        id       lon    lat   alt        name             state uf
#> 320 83936 -53.70000 -29.70 95.00 Santa Maria Rio Grande do Sul RS
#> 31  82098 -51.11667  -0.05 14.46      Macapá             Amapá AP
#>             time_zone offset_utc       time_zone.1 offset_utc.1
#> 320 America/Sao_Paulo         -3 America/Sao_Paulo           -3
#> 31      America/Belem         -3     America/Belem           -3
stns_codes <- bdmep_meta[stations_rows, "id"] 
stns_codes
#> [1] "83936" "82098"
```

Import data
-----------

Now we can import data for the two cities from 1961 to the current date.

``` r
start_date <- "01/01/1961"
end_date <- format(Sys.Date(), "%d/%m/%Y")
met_data <- bdmep_import(id = stns_codes,
                         sdate = start_date, 
                         edate = end_date, 
                         email = "your-email",
                         passwd = "your-password",
                         verbose = TRUE)
```

    #> 
    #> ------------------------------
    #> station: 83936
    #> OK (HTTP 200).
    #> 
    #> ------------------------------
    #> station: 82098
    #> OK (HTTP 200).

``` r
# check de start date
head(met_data)
#> # A tibble: 6 x 18
#>   date                id     prec  tair    tw  tmax  tmin urmax  patm
#>   <dttm>              <chr> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
#> 1 1961-01-01 00:00:00 83936    NA  NA    NA    31.9  NA      NA   NA 
#> 2 1961-01-01 12:00:00 83936    NA  23.9  21.4  NA    18.1    79  990.
#> 3 1961-01-01 18:00:00 83936    NA  30.6  25.1  NA    NA      63  988.
#> 4 1961-01-02 00:00:00 83936    NA  27.7  24.4  34    NA      75  986.
#> 5 1961-01-02 12:00:00 83936     0  26.4  23.8  NA    23.5    79  989.
#> 6 1961-01-02 18:00:00 83936    NA  30.8  25.2  NA    NA      62  989 
#> # ... with 9 more variables: pnmm <dbl>, wd <dbl>, wsmax <dbl>, n <dbl>,
#> #   cc <dbl>, evap <dbl>, ur <dbl>, ws <dbl>, request_status <chr>
```

``` r
# check de end date
tail(met_data)
#> # A tibble: 6 x 18
#>   date                id     prec  tair    tw  tmax  tmin urmax  patm
#>   <dttm>              <chr> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
#> 1 2018-07-30 12:00:00 82098  15.4  27.9    NA  NA    24.4    82 1011.
#> 2 2018-07-30 18:00:00 82098  NA    29.6    NA  NA    NA      76 1009.
#> 3 2018-07-31 00:00:00 82098  NA    27.5    NA  32.4  NA      75 1010.
#> 4 2018-07-31 12:00:00 82098   1.6  27.7    NA  NA    24      83 1012.
#> 5 2018-07-31 18:00:00 82098  NA    31.6    NA  NA    NA      69 1010.
#> 6 2018-08-01 00:00:00 82098  NA    26.6    NA  NA    NA      87 1010.
#> # ... with 9 more variables: pnmm <dbl>, wd <dbl>, wsmax <dbl>, n <dbl>,
#> #   cc <dbl>, evap <dbl>, ur <dbl>, ws <dbl>, request_status <chr>
```

You can save data in a CSV file setting argument `destdir = "path/to/write/files"` in `bdmep_import` function. Data will be save one file per station.

A description of meteorological variables can be obtained by:

``` r
bdmep_description()
#>           varname                            description  unit
#> 1            date              date and time information     -
#> 2              id                             station ID     -
#> 3            prec                          precipitation    mm
#> 4            tair                        air temperature deg C
#> 5              tw                   wet bulb temperature deg C
#> 6            tmax                maximum air temperature deg C
#> 7            tmin                minimum air temperature deg C
#> 8           urmax              maximum relative humidity     %
#> 9            patm                   atmospheric pressure   hPa
#> 10           pnmm    mean sea level atmospheric pressure   hPa
#> 11             wd                         wind direction   deg
#> 12          wsmax                              wind gust   m/s
#> 13              n                         sunshine hours     h
#> 14             cc                            cloud cover     -
#> 15           evap                            evaporation    mm
#> 16             ur                      relative humidity     %
#> 17             ws                             wind speed   m/s
#> 18 request_status Information on the status of a request  <NA>
```

Eventually, if the request failed a message will be prompted with the [HTTP status code](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes), for example:

    ------------------------
    station: 83936
    Bad Gateway (HTTP 502).
    ------------------------
    station: 82098
    Bad Gateway (HTTP 502).

In this case the outcome data frame will be filled with `NA`, except for `request_status` which will return information on the request status.

    #>   date    id prec tair tw tmax tmin urmax patm pnmm wd wsmax  n cc evap ur
    #> 1   NA 83936   NA   NA NA   NA   NA    NA   NA   NA NA    NA NA NA   NA NA
    #> 2   NA 82098   NA   NA NA   NA   NA    NA   NA   NA NA    NA NA NA   NA NA
    #>   ws          request_status
    #> 1 NA Bad Gateway (HTTP 502).
    #> 2 NA Bad Gateway (HTTP 502).


How to cite this software
---------------------

``` r
citation("inmetr")

To cite package 'inmetr' in publications use:

  Tatsch, J.D. 2018. inmetr R package (v 0.2.5): Historical Data
  from Brazilian Meteorological Stations in R. Zenodo.
  https://doi.org/10.5281/zenodo.580813.

A BibTeX entry for LaTeX users is

  @Manual{,
    title = {inmetr: Historical Data from Brazilian Meteorological Stations in R},
    author = {Jonatan Tatsch},
    year = {2018},
    note = {R package version 0.2.5},
    doi = {https://doi.org/10.5281/zenodo.580813},
    institution = {Universidade Federal de Santa Maria-UFSM},
    url = {https://github.com/lhmet/inmetr},
  }
```


NEWS.md
News
================

<!-- NEWS.md is generated from NEWS.Rmd. Please edit that file -->

# inmetr 0.3.0.9999

  - \[ \] add function to compute mean daily data.
  - \[ \] remove stations not available in BDMEP\[\] (although they
    still are in `bdmep_meta()`).
  - \[ \] check if stations in
    [climatology 1981-2010](http://www.inmet.gov.br/portal/index.php?r=clima/normaisClimatologicas)
    are available in BDMEP.

# inmetr 0.3.0

  - \[x\] fix issue \#8, thanks to @patrickgoulart12
  - \[x\] add testing for bdmep\_import()

# inmetr 0.2.8

  - \[x\] improve arguments check for `sdate` and `edate` arguments in
    `bdmep_import()`. Now, whitespace are removed from start and end of
    those arguments before building the url string to BDMEP. Thanks to
    @danielalthoff for reporting your case.

# inmetr 0.2.7

  - \[x\] fix issue related to station MACAPA - AP (OMM: 82098), thanks
    to William Barbosa (ESALQ/USP).
  - \[x\] split up bdmep\_import in `bdmep_write_csv()` and
    `bdmep_data_status()`

# inmetr 0.2.6.9999

  - \[x\] fix issue \#3, thanks to @DominickAugustine
  - \[x\] add function to write csv files of station data.
  - \[x\] include `bdmep_meta()`.

# inmetr 0.2.5

  - \[x\] include metadata of 394 meteorological stations as a dataset
    named `bdmep_meta`.
  - \[x\] Functions (see below) were renamed to have consistent names.
    Now all functions have a ‘bdmep’ prefix. This allows you to type the
    prefix and see all the members of bdmep’s family functions.
      - `import_bdmep` changes to `bdmep_import`
      - `data_description` changes to `bdmep_description`
  - \[x\] `bdmep_import()` changes:
      - supports a vector of stations IDs allowing to data import from
        multiple meteorological stations.
      - returned data frame include a new column `request status`
        (character) to inform on the outcome of the execution of the
        request on the server.

# inmetr 0.0.3

  - \[x\] fixed issue (\#1, @sillasgonzaga).

  - \[x\] `import_bmep()` has a new argument `verbose` to print if the
    status of connection is ok.


R/metadata.R
##' Metadata on 394 meteorological stations of INMET
##'
##' Spatial coordinates, altitude and others informations of INMET's
##' meteorological stations.
##'
##' @name bdmep_meta
##' @keywords datasets
##' @docType data
##'
##' @format \code{bdmep_meta} is supplied with the \code{inmetr} package and
##' consists of the following information in a data frame with 394 rows and 9 columns.
##' 	\describe{
##' 			\item{id}{station id, character with 5 digits corresponding to OMM code}
##' 			\item{lon}{longitude in decimal degrees}
##' 			\item{lat}{latitude in decimal degrees}
##' 			\item{alt}{altitude in metres}
##' 			\item{name}{station name}
##' 			\item{state}{Full name of state}
##' 			\item{uf}{Federative Unit}
##' 			\item{time_zone}{time zone}
##' 			\item{offset_utc}{UTC offset}
##' 	}
##'
##' @source \code{bdmep_meta} was compiled from
##' \url{http://www.inmet.gov.br/webcdp/climatologia/normais/imagens/normais/planilhas/Relac_Est_Meteo_NC.xls}
##' @examples
##' head(bdmep_meta)
##' with(bdmep_meta,
##'      plot(lon, lat,
##'      pch = 20,
##'      col = abs(offset_utc),
##'      main = "Met. stations of INMET")
##'      )
##' legend("bottomleft",
##'        c("UTC-5", "UTC-4", "UTC-3"),
##'        col = c(5:3),
##'        pch = 20,
##'        title = "time zone")
"bdmep_meta"


##' Description of meteorological variables
##'
##' This function describe the Meteorological variables imported by \code{\link{bdmep_import}}
##' @description Get variable names, description and units
##' @importFrom dplyr %>%
##' @details to information about instruments see \url{http://www.inmet.gov.br/portal/index.php?r=home/page&page=instrumentos}
##' @return a data frame is returned with
##'  \code{varname}, \code{description}, \code{unit}
##' @export
##' @author Jonatan Tatsch
##' @examples
##' met_vars <- bdmep_description()
##' met_vars
##'
bdmep_description <- function() {
  desc <- data.frame(
    varname = c(
      "date",
      "id",
      "prec",
      "tair",
      "tw",
      "tmax",
      "tmin",
      "urmax",
      "patm",
      "pnmm",
      "wd",
      "wsmax",
      "n",
      "cc",
      "evap",
      "ur",
      "ws"
    ),
    description = c(
      "date and time information",
      "station ID",
      "precipitation",
      "air temperature",
      "wet bulb temperature",
      "maximum air temperature",
      "minimum air temperature",
      "maximum relative humidity",
      "atmospheric pressure",
      "mean sea level atmospheric pressure",
      "wind direction",
      "wind gust",
      "sunshine hours",
      "cloud cover",
      "evaporation",
      "relative humidity",
      "wind speed"
    ),
    unit = c(
      "-",
      "-",
      "mm",
      "deg C",
      "deg C",
      "deg C",
      "deg C",
      "%",
      "hPa",
      "hPa",
      "deg",
      "m/s",
      "h",
      "-",
      "mm",
      "%",
      "m/s"
    ),
    stringsAsFactors = FALSE
  )
  new_line <- tibble::tibble(
    varname = "request_status",
    description = "Information on the status of a request",
    unit = NA_character_
  )
  desc <- dplyr::bind_rows(desc, new_line)
  return(desc)
}


##' Get coordinates on meteorological stations
##'
##' Fetch the coordinates on meteorological stations from INMET
##' \url{http://www.inmet.gov.br/projetos/rede/pesquisa/lista_estacao.php}
##'
##' @importFrom dplyr %>%
##' @return A data frame is returned with metadata, including the stations
##'  \code{id}, and coordinates (\code{lon}, \code{lat}, \code{alt})
##' @author Jonatan Tatsch

bdmep_coords <- function() {
  # omm id, lat, lon, alt
  link_stns_info <- "http://www.inmet.gov.br/sim/sonabra/index.php"

  txt <- httr::GET(link_stns_info) %>%
    httr::content("text") %>%
    textConnection() %>%
    readLines()

  closeAllConnections()
  rm(link_stns_info)

  txt_subset <-
    txt %>%
    stringi::stri_trans_general("latin-ascii") %>%
    stringr::str_subset("Codigo OMM")

  txt_info_num <- txt_subset %>%
    stringr::str_extract_all("[-+]?([0-9]*\\.[0-9]+|[0-9]+)")

  # rm(txt)

  tab_info <-
    txt_info_num %>%
    purrr::map_df(function(x) {
      # x <- a
      n <- length(x)
      x[c(3, n - (2:0))] %>%
        matrix(nrow = 1) %>%
        tibble::as_tibble() %>%
        return()
    }) %>%
    setNames(c("id", "lat", "lon", "alt")) %>%
    dplyr::select_("id", "lon", "lat", "alt") %>%
    dplyr::mutate_at(dplyr::vars(dplyr::one_of(c("lat", "lon", "alt"))), as.numeric) %>%
    data.frame()

  # tab_info <- dplyr::full_join(bdmep_stations(), tab_info, by = "id")

  return(tab_info)
} # end function






##' Get basic information on meteorological station from INMET
##'
##' This function is used to find the OMM station ID that can be
##' used to import BDMEP data using \code{\link{bdmep_import}}
##' @description Get OMM code, state and station name on meteorological stations from INMET
##' \url{http://www.inmet.gov.br/projetos/rede/pesquisa/lista_estacao.php}
##' @importFrom dplyr %>%
##' @return a data frame is returned with
##'  \code{name}, \code{state}, \code{id}
##' @author Jonatan Tatsch
##' @examples
##'
##' \dontrun{
##' # tyr get information from inmet web site
##' stns <- bdmep_stations()
##' head(stns, 15)
##' #save(stns, file = "data/stns.rda")
##' }
bdmep_stations <- function() {
  link_stns_l <- "http://www.inmet.gov.br/projetos/rede/pesquisa/lista_estacao.php"
  tab <- httr::GET(link_stns_l) %>%
    httr::content("text") %>%
    xml2::read_html() %>%
    rvest::html_node("table") %>%
    rvest::html_table(header = TRUE)

  ## remove colum named " "
  tab <- tab %>%
    names() %>%
    stringr::str_trim(.) %>%
    purrr::discard(str_empty) %>%
    subset(tab, sel = .)

  tab <- tab %>%
    names() %>%
    tolower() %>%
    # replace accented characters with non-accented counterpart
    stringi::stri_trans_general("latin-ascii") %>%
    stringr::str_replace_all(" ", "_") %>%
    stringr::str_replace("_da", "") %>%
    setNames(tab, nm = .) %>%
    tidyr::separate(nome_estacao, c("nome", "estado"), sep = " - ")
  tab <- tab %>%
    data.frame() %>%
    setNames(c("name", "state", "id")) %>%
    dplyr::mutate(id = as.character(id))
  return(tab)
}